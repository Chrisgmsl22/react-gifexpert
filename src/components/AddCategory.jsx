import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({onNewValue}) => {
    //Cada componente puede tener su propio estado
    const [inputValue, setInputValue] = useState('Dragon Ball');
    const onInputChange = ({target}) => {
        //console.log(target.value);
        setInputValue(target.value);
        //EL valor del set debe ser el valor del evento
        //setCategories(.... inputValue)
        
    }

    const onSubmit = (event) => {
        
        //Para no tener que hacer un refresh del naveador
        event.preventDefault();
        console.log(inputValue);
        //VAlidador para no introducir espacios en blanco
        if (inputValue.trim().length <=1){return;}

        
        //Solo mandamos a llamar a la funciÃ³n pasada desde el padre
        onNewValue(inputValue.trim()); //pasamos el valor limpio
        
        //Despues vaciamos el contenido del inputValue
        setInputValue('');
    }
  return (
    
        <form onSubmit={(event) => {onSubmit(event)}} aria-label='form'>
            <input 
                type="text"
                placeholder='Buscar Gifs'
                value={inputValue}
                onChange={(event) => {
                    onInputChange(event)
                }} />
        </form>
        
    
  );
}


AddCategory.propTypes = {
    onNewValue: PropTypes.func.isRequired,
}
