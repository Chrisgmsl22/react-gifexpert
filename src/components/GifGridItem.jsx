import React from 'react'
import PropTypes from 'prop-types';


export const GifGridItem = ({title, url}) => {
  
    //console.log(title, url);
    return (
    <div className='card'>
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  )
}

/**
 * AÃ±adir proptypes yarn add
 * a. title obligatorio
 * b. url obigatorio
 */

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
}
