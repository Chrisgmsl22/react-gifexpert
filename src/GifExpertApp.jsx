import {React, useState} from 'react'
import { AddCategory, GifGrid } from './components';


export const GifExpertApp = () => {

    //React Hook para mantener el estado
    const [categories, setCategories] = useState(['Dragon Ball']);

    const onAddCategory = (newCategory) => {
      console.log(newCategory);
      if(categories.includes(newCategory)) return;
      
      //setCategories(category => [...category, 'Spyro']);
      setCategories([newCategory , ...categories]);
    }

    console.log(categories);
  return (
    <>
        {/* Título */}
        <h1>GifExpertApp</h1>
        

        {/* Input (componente independiente) */}
        <AddCategory
            //setCategories={setCategories}
            onNewValue = {(event) => onAddCategory(event)}
            
            
        />
        {/* Listado de tarjetas */}
        
        <ol>
            {categories.map(category => (
                <GifGrid key={category} category={category}/>    
              ))
            }
            
        </ol>
            {/* Gif item */}
    </>
  )
}

//Creamos un functional component con un fragmento
/**
 * map() barre todos los elementos de un arreglo y regresa un elemento nuevo ¡
 * 
 * cuando se pone "on" hace referencia a que se está emitiendo un valor
 */