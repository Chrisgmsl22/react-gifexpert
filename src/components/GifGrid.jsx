import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react'
import { getGifs } from '../helpers/getGifs';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

//Aquí es donde hacemos la petición http para las API's


export const GifGrid = ({category}) => {

    const {images, isLoading} = useFetchGifs(category);
    
    //MAndamos a llamar a la función de inmediato
    
    return (
        <>
            <h3>{category}</h3>
            {
                isLoading
                ? (<h2>Cargando ...</h2>)
                : null
            }
            
            
            <div className='card-grid'>
                {
                    images.map((image) => {
                            return <GifGridItem
                                key={image.id}
                                {...image  }                               
                            />
                        }
                    )
                }
            </div>
                
            
        </>
    )
}


//Agregamos los propTypes
GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}

/**
 *  const [images, setImages] = useState([]); //Arreglo vacío
    
    const getImages = async() => {
        const newImages = await getGifs(category);
        setImages(newImages);
    }

    useEffect(() => {
        getImages();
    }, [])    
 */