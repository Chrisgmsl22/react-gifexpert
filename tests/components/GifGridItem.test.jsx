import { render, screen } from "@testing-library/react";
import { GifGridItem } from "../../src/components";

describe('Pruebas dentro del componente de gifgriditem', () => { 

   //Variables locales
   const title = 'Spyro';
   const url = 'https://www.playstation.com/es-mx/games/spyro-reignited-trilogy/'
    test('Debe de evaluar el snapshot ', () => {
       const {container} = render(<GifGridItem
         title={title} url={url}
       />);
      
       expect(container).toMatchSnapshot();
    });

    test('Debe de mostrar la imagen con el url y el alt indicado', () => { 
      render(<GifGridItem title={title} url={url} />);

      //screen.debug();
      // expect(screen.getByRole('img').src).toBe(url);
      // expect(screen.getByRole('img').alt).toBe(title);
      const {src, alt} = screen.getByRole('img');
      expect(src).toBe(url);
      expect(alt).toBe(alt);
      
     });

     test('Debe de mostrar el titulo en el componente', () => {

      render(<GifGridItem title={title} url={url} />);
      expect(screen.getByText(title)).toBeTruthy();
      
     });
     
    
 });