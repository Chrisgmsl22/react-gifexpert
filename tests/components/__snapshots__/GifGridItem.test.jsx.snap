// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas dentro del componente de gifgriditem Debe de evaluar el snapshot  1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Spyro"
      src="https://www.playstation.com/es-mx/games/spyro-reignited-trilogy/"
    />
    <p>
      Spyro
    </p>
  </div>
</div>
`;
